### PR TITLE
Allow backfilling dagruns

### PR DESCRIPTION
**Pull request by mtustin-handy**
Also, I'm using this at Handy (http://www.handy.com/careers/). handle @mtustin-handy and @marcintustin.

This does three related things:
1. Adds functionality to trigger_dag to enable it to be used to backfill dagruns in particular states. This is useful to prevent the scheduler trying to catch up with old dag runs when upgrading to version 1.6
2. Allows the scheduler to gracefully skip externally-triggered DagRuns which have names that conflict with scheduled runs. Without this, triggering a dagrun with a run_id of form "scheduled__<scheduled date>" would cause an exception in the scheduler, and prevent it making any progress if its name aligned with that of a scheduled job.
3. Also add some more logging to make it easier to understand what the scheduler is doing. 

Bonus:

Adds tests and makes the run_unit_tests.sh script idempotent by deleting the test db every time.

### CODE REVIEW

**Comment by mtustin-handy** [comment]
@mistercrunch May I commend this for your attention. I think you'll like it.

**Comment by mtustin-handy** [comment]
Updated with changes to make run_unit_tests idempotent, and adds some tests.

**Comment by mtustin-handy** [comment]
@mistercrunch Tests pass, and coverage is increased.

**Comment by mistercrunch** [comment]
Very cool, sorry about the delay.

We try to stay in line with PEP8 as much as possible and generally try to keep lines limited at around 80 chars. You can install flake8 and install git hooks that will lint only your changes everytime you commit or amend.

I'll start a round of review with inline comments.

**Comment by mtustin-handy** [comment]
@mistercrunch Gotcha, I'll go over it for style failures. Also, airflow is still probably the most responsive open source project.

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -511,10 +527,27 @@ def get_parser():
 
     ht = "Trigger a DAG"
     parser_trigger_dag = subparsers.add_parser('trigger_dag', help=ht)
-    parser_trigger_dag.add_argument("dag_id", help="The id of the dag to run")
+    parser_trigger_dag.add_argument("dag_id", help="The id of the dag to run.", default=None)
```
Comment (mistercrunch) [inline airflow/bin/cli.py:530]: `default=None` is implicit and/or not necessary

File: airflow/bin/cli.py [hunk 2]
```diff
@@ -93,22 +94,37 @@ def backfill(args):
...
-            state=State.RUNNING,
-            external_trigger=True)
-        session.add(trigger)
-        logging.info("Created {}".format(trigger))
-    session.commit()
+    try:
+        # TODO: verify dag_id
+        if args.start_date:
+            if args.run_id:
+                raise AirflowException("You may not specify a single run_id for a series of DagRuns")
```
Comment (mtustin-handy) [inline airflow/bin/cli.py]: length

File: airflow/jobs.py [hunk 3]
```diff
@@ -343,61 +343,96 @@ def import_errors(self, dagbag):
...
+                run_id=run_id,
+                execution_date=next_run_date,
+                state=State.RUNNING,
+                external_trigger=False
+            )
+            session.add(next_run)
+            session.commit()
+            return next_run
+        else:
+            logging.debug('Refusing to schedule because no next_run_date or schedule_end, or schedule_end is now: %s, %s'
```
Comment (mtustin-handy) [inline airflow/jobs.py:434]: Length

File: airflow/jobs.py [hunk 4]
```diff
@@ -343,61 +343,96 @@ def import_errors(self, dagbag):
                 filename=filename, stacktrace=stacktrace))
         session.commit()
 
+    def _active_runs(self, session, dag):
```
Comment (mistercrunch) [inline airflow/jobs.py:346]: Shouldn't this be a public dag property? You can replicate the @provide_session pattern as well

File: airflow/jobs.py [hunk 5]
```diff
@@ -343,61 +343,96 @@ def import_errors(self, dagbag):
...
+        qry = session.query(func.count()).filter(
+            DagRun.dag_id == dag.dag_id,
+            DagRun.external_trigger == False,
+            DagRun.state == State.RUNNING,
+        )
+        return qry.scalar()
+
+    def _last_scheduled_run(self, session, dag,
```
Comment (mistercrunch) [inline airflow/jobs.py:354]: Shouldn't this be a public dag property? You can replicate the @provide_session pattern as well

File: airflow/jobs.py [hunk 6]
```diff
@@ -343,61 +343,96 @@ def import_errors(self, dagbag):
...
+        Helper method for schedule_dag.
+        Given the database session, and dag, return the latest scheduled dag run
+        """
+        qry = session.query(func.max(DagRun.execution_date)).filter_by(
+            dag_id = dag.dag_id).filter(
+                    or_(DagRun.external_trigger == False,
+                        DagRun.run_id.like(scheduled_prefix)))
+        return qry.scalar()
+
+    def _next_dag_run_date(self, session, last_scheduled_run, dag):
```
Comment (mistercrunch) [inline airflow/jobs.py:366]: should be leveraging _last_scheduled_run and `dag.following_schedule`

**Comment by mistercrunch** [comment]
The refactor is nice, but I'm really afraid of touching this code since it's at the hearth of Airflow and is hard to fully test outside of production. There are many edge cases that I'm just done ironing out from last week. Things around pool, queues, @once, cron vs timedelta vs relativedelta, ...

**Comment by mistercrunch** [comment]
The ident shift made github's diffing engine throw up all over this... makes the changes around `schedule_dag` hard to review.

**Comment by mtustin-handy** [comment]
@mistercrunch I agree. The reason why I've refactored it is specifically to make it easier to read and reason about. Splitting it up also makes it easier to test sub parts of it. (I haven't actually written those tests, at least yet). My testing in our staging environment revealed the name clash as a bug. If you're fundamentally OK with this, I'll put through comments and test again in our staging.

I'll also add my company's name name to the readme in the this PR.

**Comment by mistercrunch** [comment]
Can we break this down in smaller chunks? One could be the added functionality in the CLI, and the other one the `schedule_dag` refactor. Maybe even breaking down the refactor into smaller chunks that are very easy/clear to review and that I can land at 1-2 days intervals. 

I'm normally that anal about how PRs are structure, but this is touching the heart of the system and could generate tiny bugs can create huge headaches in production.

**Comment by mtustin-handy** [comment]
@mistercrunch Alright. The refactor also includes a bug fix. Indeed, the CLI changes allow one to more easily trigger that bug. 

I'll look at splitting it out into incremental PRs which you can merge in a particular order.

**Comment by mistercrunch** [comment]
Sorry to be generating more work for you, but I don't comfortable merging this big chunk as is.

Maybe start with one PR for the bugfix and one for the CLI

I'm personally not too concerned about collision between external and scheduler triggers runids as individual DAGs will tend to be one or the other and `scheduled__` is safe.

**Comment by mistercrunch** [comment]
btw part of the reason I took a moment to take the time to review this PR is because it is fairly large and touching the heart of the system. I think most maintainers prefer smaller, targeted PRs. 3 small PRs might make it in faster than one gigantic one.

**Comment by mtustin-handy** [comment]
@mistercrunch I hear what you're saying. Look out for some more PRs today.

**Comment by mistercrunch** [comment]
Closing as from my understanding this is getting broken down

**Pull request closed without merging**